src/handlers/RebalanceHLPHandler.sol

```diff
@@ -0,0 +1,167 @@
+    sglp = _service.sglp();
+    calculator = ICalculator(_calculator);
+    minExecutionFeeBPS = _minExecutionFeeBPS;
+  }
+
+  function setWhiteListExecutor(address _executor, bool _active) external onlyOwner {
+    if (_executor == address(0)) {
+      revert RebalanceHLPHandler_AddressIsZero();
+    }
+    whitelistExecutors[_executor] = _active;
+    emit LogSetWhitelistExecutor(_executor, _active);
+  }
+
+  function setMinExecutionFeeBPS(uint16 _newExecutionFeeBPS) external onlyOwner {
+    if (_newExecutionFeeBPS == 0) {
+      revert RebalanceHLPHandler_AmountIsZero();
+    }
+    emit LogSetMinExecutionFee(minExecutionFeeBPS, _newExecutionFeeBPS);
+    minExecutionFeeBPS = _newExecutionFeeBPS;
+  }
+
+  function setRebalanceHLPService(address _newService) external nonReentrant onlyOwner {
+    if (_newService == address(0)) {
+      revert RebalanceHLPHandler_AddressIsZero();
+    }
+    emit LogSetRebalanceHLPService(address(service), _newService);
+    service = IRebalanceHLPService(_newService);
+  }
+
+  function executeLogicReinvestNonHLP(
+    IRebalanceHLPService.ExecuteReinvestParams[] calldata _params
+  ) external nonReentrant onlyWhitelisted returns (uint256 receivedGlp) {
+    if (_params.length == 0) revert RebalanceHLPHandler_ParamsIsEmpty();
+    _validateReinvestInput(_params);
+
+    // Get current HLP value
+    uint256 totalHlpValueBefore = calculator.getHLPValueE30(true);
+    // Execute logic at Service
+    receivedGlp = service.executReinvestNonHLP(_params);
+    _validateHLPValue(totalHlpValueBefore);
+  }
+
+  function executeLogicWithdrawGLP(
+    IRebalanceHLPService.ExecuteWithdrawParams[] calldata _params
+  ) external nonReentrant onlyWhitelisted returns (IRebalanceHLPService.WithdrawGLPResult[] memory result) {
+    if (_params.length == 0) revert RebalanceHLPHandler_ParamsIsEmpty();
+    _validateWithdrawInput(_params);
+    // Get current HLP value
+    uint256 totalHlpValueBefore = calculator.getHLPValueE30(true);
+    // Execute logic at Service
+    result = service.executeWithdrawGLP(_params);
+
+    _validateHLPValue(totalHlpValueBefore);
+  }
+
+  function _validateReinvestInput(IRebalanceHLPService.ExecuteReinvestParams[] calldata _params) internal {
+    // SLOAD
+    IVaultStorage _vaultStorage = vaultStorage;
+    for (uint256 i = 0; i < _params.length; ) {
+      if (_vaultStorage.totalAmount(_params[i].token) == 0) {
+        revert RebalanceHLPHandler_InvalidTokenAddress();
+      }
+      if (_params[i].amount > _vaultStorage.totalAmount(_params[i].token)) {
+        revert RebalanceHLPHandler_InvalidTokenAmount();
+      }
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function _validateWithdrawInput(IRebalanceHLPService.ExecuteWithdrawParams[] calldata _params) internal {
+    // SLOAD
+    IVaultStorage _vaultStorage = vaultStorage;
+    uint256 totalGlpAccum = 0;
+    for (uint256 i = 0; i < _params.length; ) {
+      if (_vaultStorage.totalAmount(_params[i].token) == 0) {
```

Comment:
Would this check return an error? If I tried to convert GLP into USDC when there is zero USDC.

Reply:
on us